Extended Randomize Loadout

robojumper opened the pull request:
Closes #9000.

Remaining papercuts:

* ~~Interaction with search filter isn't quite obvious~~
* ~~Existing "Randomize Loadout" button in the loadouts dropdown is unchanged~~
* There's no way to apply a randomized loadout without saving it first.
* Fashion remains TODO #8611 
* ~~Randomizing ghost but not sparrow/emblem/ship is inconsistent~~

![GIF 16 07 2023 00-01-41](https://github.com/DestinyItemManager/DIM/assets/14299449/38600eef-60b8-4f3f-9a21-b95980ad48b3)

![grafik](https://github.com/DestinyItemManager/DIM/assets/14299449/30848fb1-0dc6-481a-9755-5b1b2fcda52a)

bhollis commented on src/app/loadout-drawer/loadout-drawer-reducer.ts:
```diff
@@ -676,3 +669,163 @@ export function removeArtifactUnlock(mod: number): LoadoutUpdateFunction {
+    const isSubclass = (i: LoadoutItem) =>
+      defs.InventoryItem.get(i.hash)?.inventory?.bucketTypeHash === BucketHashes.Subclass;
+
+    const newLoadout = {
+      ...loadout,
+      items: [...loadout.items.filter((i) => !isSubclass(i)), newLoadoutItem],
+    };
```
I *think* `addItem` will do the right thing if you just try to add a new subclass - if not, it'd be helpful to have a "setSubclass" helper.

bhollis commented on src/app/loadout-drawer/loadout-drawer-reducer.ts:
```diff
@@ -676,3 +669,163 @@ export function removeArtifactUnlock(mod: number): LoadoutUpdateFunction {
+}
+
+/**
+ * Fill in items from the store's equipped items, keeping any equipped items already in the loadout in place.
```
Copy/paste comment?

bhollis commented on src/app/loadout-drawer/loadout-drawer-reducer.ts:
```diff
@@ -676,3 +669,163 @@ export function removeArtifactUnlock(mod: number): LoadoutUpdateFunction {
   });
 }
+
+/** Replace the loadout's subclass with the store's currently equipped subclass */
```
Copy/paste comment?

bhollis commented on src/app/loadout-drawer/loadout-drawer-reducer.ts:
```diff
@@ -676,3 +669,163 @@ export function removeArtifactUnlock(mod: number): LoadoutUpdateFunction {
+}
+
+/**
+ * Replace the mods in this loadout with all the mods currently on this character's equipped armor.
```
Copy/paste comment

bhollis reviewed (approved):
This is really cool! It works well with the undo function too...

Randomizing mods is particularly clever.

robojumper commented:
I changed the Randomize Loadout button in the character dropdown. The default interaction popup now allows configuring which parts of the Loadout to randomize 
![grafik](https://github.com/DestinyItemManager/DIM/assets/14299449/72efe727-b749-4aa0-8345-7986a0d7dc42)
while the alternative button just creates a random Loadout and brings it up in the Loadout drawer.

I also added ships, sparrows and emblems to the acceptable randomize types, but restricted it to items on the current store because that strikes a good balance between randomizing as much as possible and requiring tons of manual cleanup.

bhollis commented on src/app/loadout/loadout-menu/LoadoutPopup.tsx:
```diff
@@ -384,3 +334,68 @@ function filterLoadoutToEquipped(loadout: Loadout) {
+  const searchFilter = useSelector(searchFilterSelector);
+  const unlockedPlugs = useSelector(unlockedPlugSetItemsSelector(store.id));
+
+  const createRandomLoadout = useCallback(() => {
```
This could be a redux thunk-action that fishes all that selector stuff out when it's invoked rather than having to take them via `useSelector`.

robojumper replied:
Fixed, I think?

bhollis commented on src/app/loadout/loadout-menu/LoadoutPopupRandomize.tsx:
```diff
@@ -0,0 +1,139 @@
+};
+
+const checkBoxes: { label: string; d2Only?: boolean; prop: keyof RandomizeCheckboxes }[] = [
+  { label: 'Subclass', prop: 'subclass' },
```
Should this one be localized too?

bhollis commented on src/app/loadout/loadout-menu/LoadoutPopup.tsx:
```diff
@@ -115,46 +118,6 @@ export default function LoadoutPopup({
-    } catch (e) {
-      showNotification({ type: 'warning', title: t('Loadouts.Random'), body: errorMessage(e) });
-    }
-    onClick?.(e);
```
I think this line made the loadout menu close after the randomize dialog finishes - probably still want that.

bhollis commented on src/app/loadout-drawer/loadout-drawer-reducer.ts:
```diff
@@ -681,3 +662,166 @@ export function removeArtifactUnlock(mod: number): LoadoutUpdateFunction {
+      )
+    );
+
+    if (!newSubclass || !defs.isDestiny2()) {
+      return loadout;
+    }
```
This prevents randomize-subclass from working on D1. Probably you only want to skip the socketOverrides.
